src/events/provider.py

```diff
@@ -44,38 +44,45 @@ def on_subject_requested(self, event: SubjectRequestedEvent):
         if not self.charm.unit.is_leader():
             return
 
-        extra_user_roles = event.extra_user_roles or ""
-        subject = event.subject or ""
         relation = event.relation
         username = f"relation-{relation.id}"
-        password = (
-            self.charm.context.cluster.client_passwords.get(username)
-            or self.charm.workload.generate_password()
-        )
+        password = self.charm.context.cluster.client_passwords.get(username, "")
+
+        # All units can update their own authfile. If password is not yet set, wait until
+        # leader creates the password.
+        if not password and self.charm.unit.is_leader():
+            password = self.charm.workload.generate_password()
+        if not password and not self.charm.unit.is_leader():
```

Comment:
**nitpick:** I think the `and not self.charm.unit.is_leader()` is redundant here, and slightly confused me initially.

I think it reads clearer like:
- If no password, and I'm the leader, I'll make it
- Generally, if no password at this point, defer

Up to you, very minor.

Reply:
+1 Marc, it would definitely make the code leaner. But also, at line 44 there is a condition that if we are not the leader, we `return` not defer, which is a bit confusing

Reply:
Yes, I forgot to remove condition on L44. All units can execute this event handler.